Kindsys: Introduce Kind framework

sdboyer opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

-->

**What this PR does / why we need it**:

This PR introduces the kind framework to Grafana. This will be the system through which Grafana expresses canonical schemas for its objects, and will drive frontend, backend, plugin, docs, and as-code use cases through standardized tooling and codegen pipelines.

It also introduces a formal (though still evolving) maturity system for kinds. That system is intended to provide both milestones for Grafana Labs developers that are working towards converting existing systems to the kind framework, and clear guarantees to public audiences who want to depend on these canonical schemas, particularly in as-code contexts.

Kinds replace the old "coremodel" schema system (though some cleanup has been diverted to follow-up issues, so some directories remain), taking all the functionality there, generalizing it and adding a metadata layer [as outlined here](https://docs.google.com/document/d/12vYU-DRlS0yYxvXW52cOH7kSIxmJAER_4_mlA0An_QM/edit#heading=h.n59oqv9f75ux).

[This (and linked subdocs) is the official guide](https://docs.google.com/document/d/1gPzSCQn-0Qg86PYpPcR-l_kTQ4Ba1wgnNLZFKPSe8yQ/edit#heading=h.1bl5vdml7d4s) to writing schemas/kinds in tandem with this PR. While incomplete, that's still probably the most helpful context of all, because it goes into how folks are supposed to use this emerging platform/SDK.

(Note that that doc is not currently open to the public, but it will be Soon™, after we fill it in a bit more)


Deferred to follow-ups...definitely not exhaustive list
- Testing system for the modular codegen - but modularity should make feasible something pretty svelte
- Remove `pluginmeta` coremodel, and with it all trace of the name "coremodel" 🎉 #56765
- Refactor plugin codegen (`public/app/plugins/gen.go`) to use new modular codegen
- Refactor standard all existing `models.cue` forms to comply with `#SlotKind` constraints
- Actually generate some stuff outside `pkg/kinds` that hooks directly into efforts with search, object api server, etc.
- Move all slot-related stuff over to `pkg/kindsys` from `pkg/framework/coremodel`


cc @grafana/grafana-as-code @joshhunt @sakjur @IfSentient @GrafanaWriter @ryantxu @ArturWierzbicki @toddtreece

Commit:
Sketchy handwavy pass at entity meta framework

Commit:
Little nibbles

sdboyer commented:
given this:

> In our new platform architecture board discussion, it was suggested and quickly agreed to/accepted to rename "entity api/storage" with "object api/storage." I haven't followed that guidance here.

> I tried, but i simply see no way to actually create crisp, meaningful terminology when we try to take the term "object," which already carries a lot of meanings in software in general, and even more in the particular space we're in (e.g. "[object storage](https://en.wikipedia.org/wiki/Object_storage)" is a thing). I'm not saying that what's in here is perfect - but i see no path forward with "object" being the main noun.

ccing eng on that board: @torkelo @bergquist @zserge @mckn

ArturWierzbicki commented on pkg/framework/coremodel/entityframework.cue:
```diff
@@ -0,0 +1,99 @@
+#StructuredEntityType: {
+	_sharedKind
+	maturity: *"committed" | "synchronized" | "mature"
+	form: "schema"
```
```suggestion
	form: "structured"
```
?

sdboyer replied:
thank you, yes :)

ArturWierzbicki commented on pkg/framework/coremodel/entityframework.cue:
```diff
@@ -0,0 +1,99 @@
+//   - Entities are like objects - what's returned from the constructor
+//
+// There are two fundamental kinds of EntityTypes: RawEntityType and StructuredEntityType.
+#EntityType: #RawEntityType | #StructuredEntityType
```
I am struggling to understand why is it called `EntityType` rather than `EntityKind` - I thought we settled on `Kind` for the  e.g. from the doc:

> Coremodel: A collection of versioned schemas for a single “kind” of object, the logic for translating between those versions, and metadata on how they are to be used.


`type` would be reserved for grouping within a `kind`, e.g. the current datasource type

sdboyer replied:
i'm not terribly strongly attached to the use of kind vs. type here. i had written it at least three different ways before putting this PR up, sometimes including "kind" at the top level instead. i could be convinced either way. Though -

> `type` would be reserved for grouping within a `kind`, e.g. the current datasource type

This linkage raises yellow flags for me, because i think of these things in terms of schema compositions, and there is no _necessary_ limit on the number of things that might be composed.

For example, today the base datasource schema (whenever we write it) certainly composes a datasource plugin, so it makes sense to think of that as a simple parent/child typing relation.

Panels, by contrast, compose two things - query from datasource plugins, and panel from panel plugins. We conventionally talk about them in terms of the latter (we'd say `type: panel, kind: heatmap` in your wording here), but IMO there's no clear, necessary reason that's right - it would be just as correct to describe them in terms of their datasource.

There's also an interesting, not-implausible scenario for datasources as well - there are a few flavors of prometheus (e.g. vanilla vs. with sigv4) that might want us to allow _further_ differentiation of a `prometheus`-typed ds kind beyond that single level. Alternatively, we might want `type` of a datasource to refer to the query language it speaks, rather than the specific plugin.

All of these make me concerned that making a hard assumption about a 'type-subtype' relation between `kind` and `type` may bite us.

ryantxu replied:
any alternatives?  we currently just use "type" for everything.  Or rather we way "a panel of type heatmap" or "a datasource of type influx" -- to make it confusing, we also say "a plugin of type influx" and "a transformer of type joinByLabels"

ArturWierzbicki replied:
> This linkage raises yellow flags for me, because i think of these things in terms of schema compositions, and there is no necessary limit on the number of things that might be composed.

> Panels, by contrast, compose two things - query from datasource plugins, and panel from panel plugins. We conventionally talk about them in terms of the latter (we'd say kind: panel, type: heatmap in your wording here - though it's also important to note that panels are not a top-level thing in the schema system, and we aren't close to being able to make them one), but IMO there's no clear, necessary reason that's right - it would be just as correct to describe them in terms of their datasource.

Hmm... I see your general point of being able to categorize an entity in different ways, but I don't follow this example. The `type` can be useful even if the entity in question made no use of composability, and never referred to other entities. 

Imagine we had datasource concept - we would still want to differentiate between a "text" panel and a "image" panel. This isn't to say there is one-true-way of defining the `type` - we define and use it because having a clear separation between groups of entities with shared characteristics/features is useful from the application PoV.

It's a very interesting point!

sdboyer replied:
(Just to be clear, i'm not trying to change the way anyone talks about it. IMO the conversational shorthand is totally a-ok 👍🏻 excelente.)

> any alternatives?

Having a precise alternative is why i invested so much thinking into [composition and "slots"](https://docs.google.com/document/d/12vYU-DRlS0yYxvXW52cOH7kSIxmJAER_4_mlA0An_QM/edit#heading=h.knp2lhkqkje3) (a name i still hate). That should (i hope?) offer precise terms for the whole problem space. So:

The dashboard lineage has `Panel` and `Query` slots, each of which [accepts N conformant lineages](https://github.com/grafana/grafana/blob/v8.5.0/packages/grafana-schema/src/scuemata/dashboard/dist/family.cue#L30-L44). Each panel instance will be an instance of exactly one of the `Panel` and `Query` slot implementations (or maybe zero-or-one, with zero being the case where the plugin has provided no schema).

Therefore, ignoring the issue of panels not being top-level thingies for the moment, a serialized form of this info for a particular instance of a panel in a dashboard might be `panel(Panel:timeseries,Query:prometheus)`.

sdboyer replied:
> The type can be useful even if the entity in question made no use of composability, and never referred to other entities.

Sure - of course it "can be useful." But if such expansion is permitted, then the meaning/intended interpretation of the value of `type` becomes arbitrary. That's not a problem on its own - from the system's perspective, schema contents are arbitrary. That's how it should be.

But [this code](https://github.com/grafana/grafana/pull/56507/files#diff-2724413b7c4fd4e0590ba2b13c066a02d273ffc76f2e7c19bc3d87bebf10362cR111), `kind` and `type`  are being given a special, system-level function: generic components of a reference, which generic algorithms in the system are supposed to be able to do _something_ meaningful with. If `type` is just an opaque string, dependent on `kind`, it limits what any generic algorithm can really do. It also forces us to design a considerably more complex system, where instead of defining a generic algorithm, validity and backwards compatibility rules _once_, we do that _for each kind_, and it becomes the system's responsibility to try to enforce meta-rules on what's been delegated to the kinds. (Which is basically impossible, so we won't even try)

IMO, keeping a clean separation between system-level and type-level responsibilities is one of the most important things we can do to avoid [sprawl and incidental complexity](https://github.com/grafana/grafana/pull/56492#issuecomment-1272293604).

Commit:
Update pkg/framework/coremodel/entityframework.cue

Co-authored-by: Artur Wierzbicki <wierzbicki.artur.94@gmail.com>

torkelo commented:
Struggling a bit to see what this is the start of and how it's to be used, maybe start with the go types and where / how they would be used?

sdboyer commented:
> Struggling a bit to see what this is the start of and how it's to be used

Yeah, illustrating how these things will actually get used is weighing heavily on my mind. The challenge has been trying to pick something simple that doesn't involve just _doing_ all of it - which i feel like i've been circling since August.

i ended up picking this as a point of entry because the pain i'm feeling right now is centered on trying to create this schema authoring tutorial, which depends on being able to talk about stages of schema maturity, as well as referring to some clear bounding concepts (what kind of schema am i writing? what does the automation pipeline produce when i write that schema?)

> maybe start with the go types and where / how they would be used?

Maybe? It's hard for me to see a path to doing this that, again, doesn't just require doing most or all of it. But, i think i can describe it decently enough, to where you might be able to reason about the impacts of what's here.

Using the playlist type as a linked example: today, each [`coremodel.cue`](https://github.com/grafana/grafana/blob/23e04c0f9c5ee419fa54fc4bf2c9807d19f1fd51/pkg/coremodel/playlist/coremodel.cue) gets rendered to a file which defines a [`type Coremodel struct`](https://github.com/grafana/grafana/blob/23e04c0f9c5ee419fa54fc4bf2c9807d19f1fd51/pkg/coremodel/playlist/playlist_gen.go#L99-L103), which implements [`coremodel.Interface`](https://github.com/grafana/grafana/blob/23e04c0f9c5ee419fa54fc4bf2c9807d19f1fd51/pkg/framework/coremodel/interface.go#L11-L29). The generated coremodel registry [references this generated `Coremodel` type](https://github.com/grafana/grafana/blob/main/pkg/framework/coremodel/registry/registry_gen.go#L31).

After the things in this file are put into effect (reminder to please consider names as placeholders, not commitments):
* `pkg/framework/coremodel` becomes `pkg/framework/entity`
* We add the following to `pkg/framework/entity`, each of which have fields corresponding to the definition of the same name in this PR. These would be generated, or at least checked at CI time to ensure the cue and go are in sync:
  * `type RawType struct`, from `#RawEntityType`
  * `type CoreStructuredType struct`, from `#CoreStructuredEntityType`
  * `type CustomStructuredType struct`, from `#CustomStructuredEntityType`
* [`type Interface interface`](https://github.com/grafana/grafana/blob/23e04c0f9c5ee419fa54fc4bf2c9807d19f1fd51/pkg/framework/coremodel/interface.go#L11-L29) is renamed to `type Type interface`, and at least `CoreStructuredType` and `CustomStructuredType` would implement it. (More methods will be needed, but not thinking about that rn)
* The code generator is changed so that, instead of [defining](https://github.com/grafana/grafana/blob/main/pkg/coremodel/playlist/playlist_gen.go#L101-L103) `type Coremodel struct`, then [populating and returning one from `New()`](https://github.com/grafana/grafana/blob/main/pkg/coremodel/playlist/playlist_gen.go#L121-L135), an instance of `entity.CoreStructuredType` is populated and returned from `New()`.

ryantxu commented:
I will confess my cue skill are not strong enough to deeply understand the implications on how the structures are defined.  I do have pretty clear needs I can easily express in go.  See https://github.com/grafana/grafana/pull/56507 

I think most of that can be based on cue, but seems easier to have a clear target.  At least one that is real enough to test assumptions against a real SQL backing and real(ish) file objects

sdboyer commented:
i appreciate the honesty. I do understand that CUE is not most Grafana folks' preferred form of expression, and i really am trying to be cognizant of that by minimizing the amount that's actually written in it.

FWIW, my guiding instinct here is to provide Grafana with a powerful language for  declarative control of Grafana (and associated systems) behavior, while keeping those system capabilities tight enough that we don't end up with [15k-word markdown docs](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md) explaining best practices and [contextually redefining the meaning of the word "merge."](https://kubernetes.io/docs/reference/using-api/server-side-apply/#merge-strategy). It's always struck me that one of the ways things got so messy there is through the awkward way in which the definition of their basic declarative object model is sprawled out across YAML, Go, and long spec documents. Sprawl breeds compensatory complexity, which leads to more sprawl...

My impulse towards CUE here is to learn from k8s and preempt that sprawl as we formalize Grafana's object model. CUE's unique and specific skill is concisely expressing relational constraints on a universe of data, portably applying those rules at runtime or codegen-time, and all while still being reasonably amenable to static analysis, code and docs generation. Being able to functionally specify the whole core of the object model in a few files (a few! just a few!) seems to me like it could protect us from many (not all!) of the incidental complexity demons that haunt k8s-land to this day.

Commit:
Move file into new framework location

Commit:
Introduce loaders, Go code

Commit:
Complete rename to kind

Commit:
Update thema to latest

Commit:
Deal with s/Library/*Runtime/

Commit:
Commit new, working results of codegen

Commit:
We like pointers now

Commit:
Always take runtime arg for NewBase()

Commit:
Flesh out framework, add svg/dashboard examples

sdboyer commented:
I've added quite a bit of sketched-out implementation. Will make some inline comments, but there are now a bunch of the Go framework we'll need in place (even if they `panic("TODO")`), and examples to illustrate how i see this working for both raw and structured types.

There's still a major hole on how the existing coremodel base registry will be reimplemented, (and others i'm sure) but it should be at least a bit easier now to imagine where this is going.

sdboyer commented:
note that this will get considerably smaller when #56602 goes in, we needed a thema update that required some renames

sdboyer commented on pkg/framework/kind/type.cue:
```diff
@@ -0,0 +1,111 @@
+	#Structured
+}
+
+// NOTE not a definition b/c apps/plugins could be written against a later Grafana version
+CustomStructured: {
+	#Structured
+}
```
This is a bit of conventional CUE practice - [definitions](https://cuelang.org/docs/tutorials/tour/intro/schema/) (`#`-led names) are normally interpreted as closed, meaning that an error occurs if they're unified with data that has fields not specified in the definition.

This is a forward compatibility thing. So, for stuff that's expected to be used solely in Grafana core, that's fine - those things are part of the same repo, and therefore have the same/a predictable version vs. `#CoreStructured`. 

However, externally-defined plugins do not. If we introduce a new metadata field on `#CustomStructured` (with the leading hash) in, say, Grafana 10.5, and a plugin includes that field in one of their type definitions, then a running Grafana 10.4 instance would reject that plugin's type as an excess field because it's not in `#CustomStructured`.

For some cases, that may be what we want - but it should be managed explicitly with a minimum grafana version number somewhere, not implicitly through a schema metadata validation failure.

Commit:
Cruft removal

sdboyer commented on pkg/framework/kind/load.go:
```diff
@@ -0,0 +1,195 @@
+	}
+}
+
+// ParseKindFS takes an fs.FS and validates that it represents one
+// of the valid kind variants. On success, it returns a representation
+// of the entire kind definition contained in the provided kfs.
+func ParseKindFS(kfs fs.FS, rt *thema.Runtime) (*ParsedKindDefinition, error) {
+	panic("TODO")
+}
```
We already have a similar function for plugins - `pkg/plugins/pfs.ParsePluginFS`. The idea behind a func at this particular level of abstraction receiving an `fs.FS` is that we can use it:

* **at codegen-time,** when we scan the real filesystem and construct an `os.DirFS` - like [here](https://github.com/grafana/grafana/blob/main/public/app/plugins/gen.go#L60)) _and_ at runtime when loading up things in our `embed.FS` - like here
* **at runtime,** when we [load the same stuff](https://github.com/grafana/grafana/blob/f9e426cf80b83dac6abf2a22491795630deba31f/pkg/plugins/pfs/corelist/loadlist_gen.go#L35), but this time from an `embed.FS`

This is a hard-won lesson about effective code generator design. Requiring that these two different cases be treated symmetrically is a powerful design constraint: it reduces bugs and code duplication, as well as enforcing that we have a clean concept of a loader.

Commit:
Remove generated kind go files from gitignore

Commit:
Merge branch 'main' into sdboyer/entity-framework-skeleton

sdboyer commented on pkg/kind/dashboard/kind_gen.go:
```diff
@@ -0,0 +1,91 @@
+}
+
+// TODO standard generated docs
+func (k *Kind) ConvergentLineage() thema.ConvergentLineage[*Entity] {
+	return k.lin
+}
+
+// JSONValueMux is a version multiplexer that maps a []byte containing JSON data
+// at any schematized dashboard version to an instance of *Entity.
+//
+// Validation and translation errors emitted from this func will identify the
+// input bytes as "dashboard.json".
+//
+// This is a thin wrapper around Thema's [vmux.ValueMux].
+func (k *Kind) JSONValueMux(b []byte) (*Entity, thema.TranslationLacunas, error) {
+	return k.valmux(b)
+}
```
This is really the :chef-kiss: of all the new thema generics changes - a [`ConvergentLineage`](https://pkg.go.dev/github.com/grafana/thema#ConvergentLineage) encapsulates what today exists as [`coremodel.Interface`](https://github.com/grafana/grafana/blob/5641029a4afe3c368ecf6d38ee36c998083acc07/pkg/framework/coremodel/interface.go#L11-L29) into something agnostic wrt Grafana, tighter, and actually helped out by Go's type system.

It's easy to imagine this generated `JSONValueMux` method being integrated into the new entity api, as well as into the existing web framework, generically replacing the one-off validator we've written [here](https://github.com/grafana/grafana/blob/5641029a4afe3c368ecf6d38ee36c998083acc07/pkg/api/dashboard.go#L369-L380).

That said, this is also an interesting case because, [as these lines indicate](https://github.com/grafana/grafana/blob/5641029a4afe3c368ecf6d38ee36c998083acc07/pkg/api/dashboard.go#L369-L373), the dashboard schema is only expected to apply _after_ a certain schema version. We may need to adapt the code generator to allow an additional function to be injected into this value mux which performs a similar check. (Or maybe we can handle it entirely within the schema itself)

sdboyer commented on pkg/kind/dashboard/kind_gen.go:
```diff
@@ -0,0 +1,91 @@
+//go:embed kind.cue
+var cueFS embed.FS
+
+// TODO standard generated docs
+type Kind struct {
+	lin    thema.ConvergentLineage[*Entity]
+	meta   kind.CoreStructuredMeta
+	jendec vmux.Endec
+	valmux vmux.ValueMux[*Entity]
+}
```
It is expected that kind authors can attach additional handwritten methods by just adding another file in this directory. The generated registry that exposes them (successor to [this one](https://github.com/grafana/grafana/blob/ae927eab73832b37b9a7023b079a538fa0cf5a60/pkg/framework/coremodel/registry/registry_gen.go)) will expose each kind's generated `Kind` struct directly with a single method (same as [this](https://github.com/grafana/grafana/blob/ae927eab73832b37b9a7023b079a538fa0cf5a60/pkg/framework/coremodel/registry/registry_gen.go#L44)), while also exposing all of them via interfaces (same as [this](https://github.com/grafana/grafana/blob/main/pkg/framework/coremodel/registry/provide.go#L46)), so getting at the kind-specific methods will be easy for anything working directly with a particular kind.

That said, `pkg/kind` will be import-restricted via ruleguard to disallow importing most things inside `pkg` (at minimum, nothing under `pkg/service`), in the interest of keeping this entire subsystem light and externally importable. The effect should be that stateless, storage- and api-agnostic logic can be declared here, and other stuff will need to be layered on in other packages, analogous to the layering we've already done [with the playlist service](https://github.com/grafana/grafana/blob/main/pkg/services/playlist/model.go#L25-L27).

Commit:
Refine maturity concept, add SlotKind

IfSentient commented on pkg/framework/kind/type.cue:
```diff
@@ -0,0 +1,140 @@
+}
+
+// TODO
+CustomStructured: {
```
Should this be a definition?

sdboyer replied:
i actually just removed this comment, heh

```
// NOTE not a definition b/c apps/plugins could be written against a later Grafana version
```

sdboyer replied:
i think i'll switch it back to a definition, but explicitly marked open (`...`)

Commit:
Update embed and go deps

Commit:
Export PrefixWithGrafanaCUE

Commit:
Make the loader actually work, holy crap

Commit:
Many small tweaks to type.cue

ryantxu commented on pkg/framework/kind/kind.go:
```diff
@@ -0,0 +1,49 @@
+	Name() string
+
+	// TODO docs
+	Maturity() Maturity // TODO unclear if we want maturity for raw kinds
```
Is this where you imagine having access to `models.ObjectSummaryBuilder`?

https://github.com/grafana/grafana/blob/main/pkg/services/store/kind/registry.go#L21

sdboyer replied:
Probably yes - in general, capabilities that are declared on the CUE kind framework side should have corresponding things in these interfaces (when they require a computation/behavior) or on the various `Meta` structs (when they're really just a simple config value)

The only reason it's "probably" is because, as i noted [here](https://github.com/grafana/grafana/pull/56492/files/f40576ff785baef118e222e7ff918d925bd9e42e#r992287075), we have to keep the basic generated kinds separated from the bulk of Grafana dependencies. For kinds with a go passthrough func to call for the object summary builder, i suspect that will have dependencies that violate that rule. 

The solution is basically just more layers/veneering - it's perfectly fine for us to generate the dependency-needing code under `pkg/service` in some disciplined way. Maybe that'll also mean a separate interface, too, that layers on top of these. It may even make the most sense to just handle all the object summary production outside of `pkg/kind`, as that's not something which we expect to be useful in any as-code cases, anyway.

Commit:
Add Apache 2 licensing exceptions for kinds

Commit:
Add new kinds dir, start of generator

Commit:
Roll back to earlier oapi-codegen

Commit:
Introduce new grafana-specific CUE loaders

Commit:
Introduce new tidy code generators framework

Commit:
Catch up kind framework with tinkering

Commit:
Add slices for the generators

Commit:
Add write/verify step to main generator

Commit:
Many renames

Commit:
Split up kind framework cue files

Commit:
Use kind.Decl within generated kinds

Commit:
Create kind.SomeDecl wrapper type to cache lineages

Commit:
Better names again

Commit:
Get one generated implemented, hopefully

Commit:
Copy dashboard schema into new kind.cue

Commit:
Small fixes to make the initial gen work

Commit:
Put svg kind in its new home

Commit:
Add generated Go dashboard type

sdboyer commented:
I think this PR is over the hump, and is now on the downhill to completion! Fair bit of work still to go, but the essential architecture is in place.

Key highlights:
* There's a raw kind declaration [for svg](https://github.com/grafana/grafana/pull/56492/files#diff-1f92e323ed6d2a212bfc673d13dd2b06aed89262ad40e32b78f14866de8dd65c)
* There's a core structured kind declaration [for dashboard](https://github.com/grafana/grafana/pull/56492/files#diff-eaa2367060e13c676eaa9f2c883f3bbc90e0ab1bd6aacabd468cd0e3769f3790)
* Both are declared under `kinds/`, which is a new top-level dir that's owned by _neither_ the frontend or backend
* Taking advantage of CUE's hierarchical loading, we need express only time in a parent dir that [`kinds/structured/*` is for `CoreStructured` kinds](https://github.com/grafana/grafana/pull/56492/files#diff-9daf2a8b9790c9c76853a9a9a7fc96d189d71d51dcedc017c073a84da907df5f), and [`kinds/raw/*` is for raw kinds](https://github.com/grafana/grafana/pull/56492/files#diff-5a0427e2b6f0bcb0830ccbd20062b643d60e1f0ff4df86eb1773059d22e9b504).
* The [new generator](https://github.com/grafana/grafana/pull/56492/files#diff-2e1fa6698fec2f3c845997e7986cb4c2d2d23a6234683adefee5d5db5c8af9f3) is correctly producing [generated Go types output for dashboards](https://github.com/grafana/grafana/pull/56492/files#diff-d6757d11e62460735ce4ae851d955bb43ff0955eaa7f06604894cee9ac8205f9), and correctly NOT producing any such output for svg.

It's worth going into some detail about the new way the generators are written, because i expect it to be possible for intrepid engineers to be able to add to the kind framework and codegen system. i'll do that in a follow-up comment.

sdboyer commented:
The new core kind generator in this PR (and eventually, i expect, all kind generators), are now modular. Modularizing is worthwhile because we know we have at least the following cases:

* Core kinds (this one)
* Slot kinds (the still-not-great name for what's in `models.cue` today)
* Custom kinds (at minimum in in [operator-app-sdk](https://github.com/grafana/operator-app-sdk)
* Now, multiply the above in every as-code context/ecosystem we want to target (Jsonnet, Go, CUE, TS, Terraform, Python, etc...).

Prior to this PR, it seemed generally plausible that we could reuse the same codegen system for all of the above. Now - on the fourth iteration - i am as close to certain as i get that we can do it, and that this is how it looks.

With modular codegen, there's really three parts to a top-level `func main` code generator:

* Load and validate the relevant CUE from disk representing our kinds, resulting in a [`kind.Decl`s](https://github.com/grafana/grafana/pull/56492/files#diff-dfd03d0e59d7d52d464521385791316f3fc799b078107f401a313ec31a094d3dR188-R197). That's what [these blocks](https://github.com/grafana/grafana/pull/56492/files#diff-2e1fa6698fec2f3c845997e7986cb4c2d2d23a6234683adefee5d5db5c8af9f3R50-R88) do, and it's basically all path-juggling - the hard work is done by standardized library functions.
* Run these collected `kind.Decl`s through [a set of modular code generators](https://github.com/grafana/grafana/pull/56492/files#diff-2e1fa6698fec2f3c845997e7986cb4c2d2d23a6234683adefee5d5db5c8af9f3R22-R28), each of which take either one or all collected `kind.Decl`s, and generate exactly one file.
* [Write (or verify)](https://github.com/grafana/grafana/pull/56492/files#diff-2e1fa6698fec2f3c845997e7986cb4c2d2d23a6234683adefee5d5db5c8af9f3R117-R127) the resulting generated files to disk

To date, other Grafana code generators have been _kinda_ trending in this direction. But what's here really nails it down.

`kind.Decl` is the lynchpin - it encompasses the variation across kind categories and offers nearly all the info that the modular generators need. At the same time, it's rigorously defined enough that the main generator has to do little more than shuffle some paths around to produce them.

All of this means that the loop for extending the kind framework and (when needed) its supporting codegen is basically:

1. Add some new field(s) to the family tree of types under `#Kind` in the .cue files in `pkg/framework/kind` for the use case, and make field additions [to the corresponding Go structs](https://github.com/grafana/grafana/pull/56492/files#diff-a01d3679dff33a05e1e9ca2fa3122f611e383bfea398580d2f6d119c48c0cc18)
2. Write a new implementation of either [KindGenStep](https://github.com/grafana/grafana/pull/56492/files#diff-6960112f1b3338c7fec3a78342b03dccf8ce988a7807134080ad0cc314bd70e0R14-R20) or [AggregateKindGenStep](https://github.com/grafana/grafana/pull/56492/files#diff-6960112f1b3338c7fec3a78342b03dccf8ce988a7807134080ad0cc314bd70e0R28-R34), learning from the patterns in the other ones already written
3. Add your new generator to the generator list in the relevant main generator
4. Run `make gen-cue` to test it out, tweak until the output looks right
5. Profit!

sdboyer commented:
Concrete example i discussed with @sakjur earlier today: let's say that, rather than handwriting these playlist aliases:

https://github.com/grafana/grafana/blob/ed1176adc8f47a99984cc6d514d59034511c8602/pkg/services/playlist/model.go#L25-L27

We want to generate them, inside `pkg/services/playlist`. The benefit here isn't the conventional codegen value of boilerplate reduction or saving time, but correctness and consistency: it would be a major step towards enforcing that the backend service associated with a kind is in sync with the schema/the frontend.

(Note that this is something we can and should do immediately after this PR is merged)

Applying the steps:

### Add fields & constraints to the kind framework

We could add a the following to `#CoreStructured`:

```cue
#CoreStructured: {
  // ... omitting existing fields

  // linkServiceTypes determines whether type aliases pointing to canonical 
  // generated types should be generated in the kind's backend service.
  //
  // Always true for kinds of at least stable maturity.
  linkServiceTypes: bool | *false
  if (maturity & ("experimental" | "stable" | "mature") != _|_ {
    linkServiceTypes: true
  }
}
```

We've now expressed and documented the field, as well as making it part of a maturity gate - we can know _a priori_ that any core structured type of at least `stable` maturity will have this generated type.

All that needs to be added [on the Go side](https://github.com/grafana/grafana/pull/56492/files#diff-a01d3679dff33a05e1e9ca2fa3122f611e383bfea398580d2f6d119c48c0cc18R15-R19) is this field:

```go
type CoreStructuredMeta struct {
    // ...
    LinkServiceTypes bool `json:"linkServiceTypes"`
}
```

### Write a generator

This'd be at least implicitly dependent on [the `GoTypesGenerator`](https://github.com/grafana/grafana/pull/56492/files#diff-5031889e50c1787200c48ea3196d9a8638d46b02676ee589e89b02ddff755339R72-R120) - it needs to reference code output by that generator. But the actual code to be generated is tiny - this template should be mostly sufficient:

```
package {{ .PackageName }}

import "github.com/grafana/grafana/pkg/kind/{{ .KindName }}"

{{ range .Types }}
type {{ .TypeName }} = {{ .KindName }}.{{ .TypeName}}{{ end }}

```

### Add your generator to the lists, run `make gen-cue`, profit...

Yup.

The hard part here isn't writing the code - it's figuring out what invariants we want to enforce, and how to easily express them. (One sweet spot, [as we've seen](https://github.com/grafana/grafana/pull/54816#issuecomment-1256341790), is using codegen to express the invariant in a way that gets the compiler to effectively enforce it for us. That applies here.) Once we've worked that out, the modular codegen and kind framework make it near-trivial to express

Commit:
More renames and cleanups

Commit:
Add base kind registry and generator

sdboyer commented:
@ryantxu [latest commit](https://github.com/grafana/grafana/commit/a6eed258a92e99f9a3688615a8c7e0c77f8d0575) introduces the new kind registry - the one that's safely externally importable.

It should now be easier to think about another one layered on top for search's purposes - or just something as simple as a generated slice or two with all the names, if that's what you need

Commit:
Stop blacklisting *_gen.go files

This is not the Go best practice, anyway. All we actually want to ignore
for enterprise is generated wire files.

Commit:
Change codegen output directories

pkg/kind -> pkg/kinds
pkg/registry/kindreg -> pkg/registry/corekind

Commit:
Rename pkg/framework/kind to pkg/kindsys

Commit:
Add core structured kind generator

Commit:
Add plural and machine names to kind spec

sdboyer commented:
Latest commit introduces a set of related name fields, defaults, and constraints between them - pascal case and plural. So that means, from saying

```cue
name: "SVG"
```

we end up with:

```
    Name:              "SVG",
    PluralName:        "SVGs",
    MachineName:       "svg",
    PluralMachineName: "svgs",
```

I actually am trying to keep this size of this PR minimal - the only reason i'm including this is because having the _right_ name field has a big impact on how we write code generators as well as docs.

Commit:
Copy playlist over to kind system

Commit:
Consolidate kindsys files

Commit:
Add raw kind generator

Commit:
Merge remote-tracking branch 'origin/main' into sdboyer/entity-framework-skeleton

Commit:
Update CODEOWNERS for kind framework

Commit:
Touch up comments a bit

Commit:
More docs tweaks

Commit:
Remove generated types to reduce noise for review

ryantxu commented on kinds/structured/playlist/playlist_kind.cue:
```diff
@@ -0,0 +1,47 @@
+
+lineage: seqs: [
+	{
+		schemas: [
```
are we *moving* the cue files here?  or is this another copy?  -- presumably the coremodel one goes away

sdboyer replied:
yes, they'll move, `pkg/coremodel` will be removed, and we will never see that damn word again 🎉

ryantxu commented on pkg/kindsys/kind.cue:
```diff
@@ -0,0 +1,106 @@
+//   - Byte sequences are like arguments to the class constructor
+//   - Entities are like objects - what's returned from the constructor
+//
+// There are four essential categories of kinds: Raw, Slot, CoreStructured and CustomStructured.
```
slot vs custom structured 🤔 

external plugins will be both?

sdboyer replied:
Any plugin could offer both, though we might not want core plugins to do it.

The wordsmithing for `#Kind` is...really hard 😭

sdboyer commented on pkg/kindsys/slot.cue:
```diff
@@ -0,0 +1,17 @@
+// describe the structure of its queries, which is then composed into dashboards
+// and alerting rules.
+//
+// Each SlotKind is an implementation of exactly one Slot, a shared meta-schema
```
ahh yes, thank you for the reminder - that's [here](https://github.com/grafana/grafana/blob/main/pkg/framework/coremodel/slots.cue). It's been in for a while, went in with the pfs stuff. i just haven't moved it from `pkg/framework/coremodel` to `pkg/kindsys` yet.

There will be some handwaving here, because i haven't yet defined the exact way we enforce the slot-defined meta-schema on a composed lineage in thema itself. That's grafana/thema#8 and grafana/thema#76.

In any case, [the planned architecture](https://docs.google.com/document/d/12vYU-DRlS0yYxvXW52cOH7kSIxmJAER_4_mlA0An_QM/edit#heading=h.m040j2ma3lf) for this has not changed since i wrote it.

sdboyer replied:
i was thinking i could move the file over in this PR, but that would plausibly pull in a bunch of changes to pfs/pluginmeta as well. hmm.

yeah, i think it's best to leave it where it is for right now, then move it over as part of finishing #56765. Reason being that what that PR specifically needs is another main code generator, and i don't want to write another of those without the modular farmework being introduced here.

sdboyer commented on pkg/kindsys/customstructured.cue:
```diff
@@ -0,0 +1,9 @@
+#CustomStructured: {
+	#Structured
+	maturity: *"experimental" | "mature"
+	...
```
This PR has done very little with `#CustomStructured`. i do expect it'll be _mostly_ same as `#CoreStructured`. That's borne out by the fact that most of the work in this PR has been in either `#Structured` or `_sharedKind` - and i actually did initially try to have them share a definition.

But we a number of known cases already where the constraints are different. [This example](https://github.com/grafana/grafana/pull/56492#issuecomment-1284791083) is just one.

Once i accepted that and leaned in, i found that designing for this had very nice benefits for the Go API - for example, funcs can clearly restrict their capabilities simply by [the set of metas they accept in their generic type parameter](https://github.com/grafana/grafana/pull/56492/files#diff-99bfecfb6b28f759f4e21b4931e140c1e3aa97a158b4fdcbab4c540181effc95R248).

sdboyer replied:
> It will have a fixed plugin/owner prefix? eg core can define "dashboard" but custom must be "k6-xyz"?

They will need namespacing at some level. That ties in with [the discussion](https://raintank-corp.slack.com/archives/C03BH67DS2J/p1666279859347609) @IfSentient and i were having today about how to handle "group" in k8s.

But i suspect we'll be able to express that directly in the schema when we get around to adding these as a field like this to the `pluginmeta` lineage:

```cue
  	// Unique name of the plugin. If the plugin is published on
	// grafana.com, then the plugin id has to follow the naming
	// conventions.
	id: string & strings.MinRunes(1)

	// ... other pluginmeta fields

	customTypes?: [string]: kindsys.#CustomStructured
```

Having `id` available means we can use it to establish a constraint relationship for all the `kindsys.#CustomStructured` values in `customTypes`. Same idea as what [this PR does with `machineName`](https://github.com/grafana/grafana/pull/56492/files#diff-fb814e775a368ffb53b5156eaa4d5a252647962b0017e3ab77bf4050526bc1d3R47)).

Maybe the constraints go on the custom type's `name` or maybe we make a new field. Doesn't really matter - as long as we're letting `plugin.cue` be an optional drop-in replacement for `plugin.json`, then the world is our oyster.

sdboyer replied:
> It looks like the difference here is around default maturity

The difference is both the default, and the allowable maturity stages:

* `#CoreStructured` allows `"committed" | "experimental" | "stable" | "mature"`, with `"committed"` as default
* `#CustomStructured` allows `"experimental" | "mature"`, with `"experimental"` as default

This is a function of trying to give [some clear meaning to maturity](https://docs.google.com/document/d/1gPzSCQn-0Qg86PYpPcR-l_kTQ4Ba1wgnNLZFKPSe8yQ/edit#heading=h.ajykcvt65mfb) - including public-facing guarantees - because only then can maturity milestones meaningfully reduce risk.

What's in this PR reflects what's in the doc - as [the table](https://docs.google.com/document/d/1gPzSCQn-0Qg86PYpPcR-l_kTQ4Ba1wgnNLZFKPSe8yQ/edit#heading=h.1bl5vdml7d4s) indicates, `committed` and `stable` are only currently planned to have meaning for core structured types. The only sane default i can think of is having unspecified be the first one for each kind category.

Commit:
Split each generator into its own file

Commit:
Rename Slot kind to Composable kind

Commit:
Add handwavy types for customkind loading

sdboyer commented on pkg/kindsys/load.go:
```diff
@@ -0,0 +1,234 @@
+	return decl, nil
+}
+
+// LoadCustomKind takes an fs.FS that contains a CustomKind declaration at the
+// path indicated by dir, and validates that it contains a valid CustomKind
+// declaration. On success, it returns a Go representation of the entire
+// CustomKind declaration.
+func LoadCustomKind(kfs fs.FS, dir string, ctx *cue.Context) (*Decl[CustomStructuredMeta], error) {
+	inst, err := cuectx.LoadInstanceWithGrafana(kfs, dir)
+	panic("TODO")
+}
+
+// NOTE exists solely for discussion purposes
+func LoadCustomKind2(kfs fs.FS, dir string, ctx *cue.Context) (*CustomKind, error) {
+	panic("JUST SPITBALLING")
+}
+
+// NOTE exists solely for discussion purposes
+type CustomKind struct {
+	Decl    *Decl[CustomStructuredMeta]
+	Lineage thema.Lineage
+}
```
@IfSentient i added these temporarily solely to provoke some conversation with you - `LoadCustomKind` and `LoadCustomKind2` here are both forms i could easily imagine adding, if they're something you could reasonably layer on top of. Won't be this PR, but i know you're thinking about this right now.

`CustomKind` would be filled out to become a generic version of what we're generating per-kind under `pkg/kinds/*/*_kind_gen.go`

 of these forms preferable? Or are

sdboyer commented:
i've changed the naming for "Slot Kinds" to "Composable Kinds," which is a mouthful but not wrong, and (i think) at least sorta indicative of what's going on with it.

Commit:
Guard against init calls to framework loader

Commit:
First pass at doc on extending the kind system

sdboyer commented on pkg/kindsys/EXTENDING.md:
```diff
@@ -0,0 +1,60 @@
+# Kind System
```
Threw down a quick first pass at a guide to extending the kind system. Something more durable than comments in this PR, and reflects how i expect maintenance of this to work.

This lacks [the concrete example i gave in this comment](https://github.com/grafana/grafana/pull/56492#issuecomment-1284791083), but that can be added later.

probably most relevant for @sakjur @ryantxu @ArturWierzbicki @papagian @mckn @IfSentient   @joanlopez @xlson

sdboyer replied:
oh and @wbrowne, this doc starts to cover what we talked about the other day

mckn replied:
Super! Thanks @sdboyer 🙏🏻

sdboyer replied:
ahh and @yuri-tceretian! this is what you'd use to add a thing for that contextual logging capability we discussed in backend arch meeting some months back

Commit:
Improve attribute example in docs

Commit:
Fix wire imports

Commit:
Add basic TS types generator

Commit:
Fix composable kind category def

Commit:
No need for a separate file with generate directive

Commit:
Catch dashboard schema up

Commit:
Rename generator types to something saner and generic

Commit:
Make version configurable in ts/go generators

Commit:
Add CommonMeta to ease property access

Commit:
Add kindsys prop indicating whether lineage is group

Commit:
Put all kind categories back in a single file

Commit:
Finish with kindsys group props

Commit:
Refactor maturity progression per discussion

- Replace "committed" with "merged"
- All kindcats can use all maturity levels, at least for now

Commit:
Convert ts veneer index generator to modular system

Commit:
Move over to new jennywrites framework

Commit:
Strip down old coremodel generator

Commit:
Use public version of jennywrites

Commit:
Pull latest thema

Commit:
Commit generated Go types

Commit:
Add header injection postprocessor

Commit:
Move sdboyer/jennywrites to grafana/codejen

Commit:
Merge remote-tracking branch 'origin/main' into sdboyer/entity-framework-skeleton

sdboyer commented:
All basic goals are complete.

I'm gonna now do what cleanup i can, including removing existing dashboard and playlist coremodel. And obviously make sure tests are passing, but i think this is essentially ready. (With lots of follow-ups, at least some of which are listed here https://github.com/grafana/schematization-and-as-code-project/issues/22)

Commit:
Tweak header output

Commit:
Remove dashboard and playlist coremodels

Commit:
Fix up backend dashboards devenv test

Commit:
Fix TS import patterns to new gen filename

Commit:
Update internal imports, remove coremodel registry

Commit:
Fix compilation errors, wire generation

Commit:
Export and replace the prefix dropper

Commit:
More Go struct and field name changes

Commit:
Last name fixes, hopefully

Commit:
Fix lint errors

Commit:
Last lint error

Commit:
Merge remote-tracking branch 'origin/main' into sdboyer/entity-framework-skeleton

ryantxu reviewed (approved):
This is a HUGE step forward -- looks really good.
